### PR TITLE
add jpegdec/caps:dc

### PR DESCRIPTION
**Pull request by HookedBehemoth**

### CODE REVIEW

**Comment by HookedBehemoth** [comment]
The names for the options are from libjpegs "do_fancy_upsampling" and "do_block_smoothing".

File: nx/include/switch/services/caps.h [hunk 1]
```diff
@@ -65,7 +65,10 @@ typedef struct {
 
 /// ScreenShotDecodeOption
 typedef struct {
-    u8 unk_x0[0x20];                            ///< Unknown. Set to all-zero by official sw.
+    bool fancy_upsampling;
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: u8s

File: nx/include/switch/services/capsdc.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @file capsdc.h
+ * @brief Jpeg Decoder (caps:dc) service IPC wrapper. Only Available on 4.0.0+.
```
Comment (yellows8) [inline nx/include/switch/services/capsdc.h]: `Only available on [4.0.0+].`

File: nx/include/switch/services/capsdc.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * @brief Decodes a jpeg buffer into RGBX.
+ * @param[in] width Image width.
+ * @param[in] height Image height.
+ * @param[in] opts CapsScreenShotDecodeOption decode options.
+ * @param[in] jpeg Jpeg image input buffer.
+ * @param[in] jpeg_size Input image buffer size.
+ * @param[out] image RGBA8 image output buffer.
+ * @param[in] image_size Output image buffer size, should be at least large enough for RGBA8 width x height.
+ */
+Result capsdcDecodeJpeg(const u32 width, const u32 height, const CapsScreenShotDecodeOption *opts, const void* jpeg, const u64 jpeg_size, void* image, const u64 image_size);
```
Comment (yellows8) [inline nx/include/switch/services/capsdc.h]: width/height aren't ptrs, so const isn't needed.

File: nx/include/switch/services/caps.h (hunk 1 above)
Comment (yellows8) [reply nx/include/switch/services/caps.h]: Meant bool->u8, for both of these.

**Comment by HookedBehemoth** [comment]
Actually no... What was I thinking...
The options need to have a u64 so libstrat get's the order right.

File: nx/include/switch/services/capsdc.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * @brief Decodes a jpeg buffer into RGBX.
+ * @param[in] width Image width.
+ * @param[in] height Image height.
+ * @param[in] opts CapsScreenShotDecodeOption decode options.
+ * @param[in] jpeg Jpeg image input buffer.
+ * @param[in] jpeg_size Input image buffer size.
+ * @param[out] image RGBA8 image output buffer.
+ * @param[in] image_size Output image buffer size, should be at least large enough for RGBA8 width x height.
+ */
+Result capsdcDecodeJpeg(u32 width, u32 height, const CapsScreenShotDecodeOption *opts, const void* jpeg, const u64 jpeg_size, void* image, const u64 image_size);
```
Comment (fincs) [inline nx/include/switch/services/capsdc.h]: - `size_t` for jpeg_size and image_size
- Rename image/image_size to out_image/out_image_size to make it clear that this is output.
- Note that const integer types as function arguments do nothing differently from non-const integer types.

File: nx/source/services/capsdc.c [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+void _capsdcCleanup(void) {
+    serviceClose(&g_capsdcSrv);
+}
+
+Service* capsdcGetServiceSession(void) {
+    return &g_capsdcSrv;
+}
+
+Result capsdcDecodeJpeg(u32 width, u32 height, const CapsScreenShotDecodeOption *opts, const void* jpeg, const u64 jpeg_size, void* image, const u64 image_size) {
```
Comment (fincs) [inline nx/source/services/capsdc.c]: Same changes here

File: nx/include/switch/services/caps.h (hunk 1 above)
Comment (fincs) [reply nx/include/switch/services/caps.h]: This wasn't fixed? ping @yellows8

Comment (yellows8) [reply nx/include/switch/services/caps.h]: Yeah not fixed.

Comment (HookedBehemoth) [reply nx/include/switch/services/caps.h]: Oh... missunderstood what you meant

File: nx/include/switch/services/caps.h [hunk 6]
```diff
@@ -65,7 +65,10 @@ typedef struct {
 
 /// ScreenShotDecodeOption
 typedef struct {
-    u8 unk_x0[0x20];                            ///< Unknown. Set to all-zero by official sw.
+    u8 fancy_upsampling;    ///< See libjpeg-turbo do_fancy_upsampling.
+    u8 block_smoothing;     ///< See libjpec-turbo do_block_smoothing.
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: typo

File: nx/include/switch/services/capsdc.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+void capsdcExit(void);
+
+/// Gets the Service for caps:dc.
+Service* capsdcGetServiceSession(void);
+
+/**
+ * @brief Decodes a jpeg buffer into RGBX.
+ * @param[in] width Image width.
+ * @param[in] height Image height.
+ * @param[in] opts CapsScreenShotDecodeOption decode options.
```
Comment (yellows8) [inline nx/include/switch/services/capsdc.h]: \ref CapsScreenShotDecodeOption

File: nx/include/switch/services/capsdc.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @file capsdc.h
+ * @brief Jpeg Decoder (caps:dc) service IPC wrapper. Only Available on [4.0.0+].
```
Comment (yellows8) [inline nx/include/switch/services/capsdc.h]: As previously mentioned, lowercase `available`.

File: nx/source/services/capsdc.c [hunk 9]
```diff
@@ -0,0 +1,37 @@
+#define NX_SERVICE_ASSUME_NON_DOMAIN
+#include "service_guard.h"
+#include "services/capsdc.h"
+
+static Service g_capsdcSrv;
+
+NX_GENERATE_SERVICE_GUARD(capsdc);
+
+Result _capsdcInitialize(void) {
+    return smGetService(&g_capsdcSrv, "caps:dc");
```
Comment (yellows8) [inline nx/source/services/capsdc.c]: Add a sysver check.

**Pull request merged**